src/benchmarks/micro/runtime/Interop/ComWrappers.cs

```diff
@@ -0,0 +1,118 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using BenchmarkDotNet.Attributes;
+using MicroBenchmarks;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Interop
+{
+    [BenchmarkCategory(Categories.Runtime)]
+    public unsafe class ComWrappersTests
+    {
+        static readonly ComWrappers s_instance = new DummyComWrappers();
+
+        static object QueryRCW(nint ptr)
+            => s_instance.GetOrCreateObjectForComInstance(ptr, CreateObjectFlags.None);
+
+        private List<nint> iterations;
+        private object targetRcw;
+
+        public ComWrappersTests()
+        {
+            // Allocate a simple IUnknown instance.
+            var ptr = (void**)NativeMemory.Alloc(5, (nuint)sizeof(void*));
+
+            ptr[1] = (void*)1; // Initial ref count.
+
+            // Define IUnknown vtable.
+            ptr[2] = (delegate* unmanaged<void*, Guid*, void**, int>)&QueryInterface;
+            ptr[3] = (delegate* unmanaged<void*, int>)&AddRef;
+            ptr[4] = (delegate* unmanaged<void*, int>)&Release;
+
+            // Assign vtable to instance.
+            ptr[0] = &ptr[2];
+
+            // Record RCW pointer.
+            nint rcwPtr = (nint)ptr;
+
+            // Transfer ownership to RCW instance. This populates the RCW cache.
+            targetRcw = s_instance.GetOrCreateObjectForComInstance((nint)ptr, CreateObjectFlags.None);
+
+            // Ownership was transferred.
+            Marshal.Release((nint)ptr);
+
+            // Define a large number of iterations for parallel action.
+            var limit = 3_000_000;
+            iterations = Enumerable.Repeat(rcwPtr, limit).ToList();
+        }
+
+        [Benchmark]
+        public unsafe ParallelLoopResult ParallelRCWLookUp()
+            => Parallel.ForEach(iterations, static (nint ptr) => QueryRCW(ptr));
```

Comment:
I know that I have suggested `Parallel.ForEach` when we talked about this. I do not see this pattern used for testing concurrent anywhere in this repro. I am not sure how well it is going to work (e.g. how stable it is going to be).

I would stick to one of the existing concurrent execution patterns used in the repo. For example, this one:

https://github.com/dotnet/performance/blob/c30478de1d1e8903221d8b21920a9c980d65ee55/src/benchmarks/micro/libraries/System.Buffers/ArrayPoolTests.cs#L63-L81

Reply:
Just curious, what does Parallel.For do that is meaningfully different to what is done there? In both cases it's presumably one threadpool thread per core.

Reply:
`Parallel.For` takes IEnumerable as the input. It needs to take lock when fetching and scheduling items from this enumerable. 

The work being tested in each iteration is very small: take a lock and do a lookup in a dictionary. The lock taken by `Parallel.For` scheduler can dominate or otherwise interfere with the work being tested.